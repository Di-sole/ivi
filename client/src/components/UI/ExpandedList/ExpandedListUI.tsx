import { ReactNode } from 'react'
import styles from './ExpandedListUI.module.scss'
import Link from 'next/link';

interface ExpandedListUIProps {
    children: ReactNode;
    toggleButtonTexts: string[];
    clipSize?: number;
    clipFormat?: string;
}

function ExpandedListUI({ children, toggleButtonTexts, clipSize = 2, clipFormat = 'vertical' }: ExpandedListUIProps) {

    function handleShowText({ currentTarget }: React.MouseEvent<HTMLSpanElement>) {
        const text = document.getElementById('text-wrapper');
        if (text && text.classList.contains(styles['is-truncated'])) {
            text.classList.remove(styles['is-truncated']);
            currentTarget.innerHTML = toggleButtonTexts[1];
        } else {
            text?.classList.add(styles['is-truncated']);
            currentTarget.innerHTML = toggleButtonTexts[0];
        }
    }

    return (
        <>
            <div className={styles.clause} role='expanded-wrapper'>
                <div
                    id='text-wrapper'
                    data-testid='clipped-text'
                    className={[styles['clause-text'], styles['is-truncated']].join(' ')}
                >
                    <div className={[styles['clause-text-inner'], styles['hidden-children'], 'clause-clamped'].join(' ') }>
                        {children}
                    </div>
                    <span
                        className={styles['clause-toggle']}
                        onClick={handleShowText}>
                        {toggleButtonTexts[0]}
                    </span>
                </div>
            </div>
            <style jsx>
                {`
                    .clause-clamped {
                        -webkit-line-clamp: ${clipSize};
                        -webkit-box-orient: ${clipFormat};
                    }
                `}
            </style>
        </>
    )
}

export default ExpandedListUI;