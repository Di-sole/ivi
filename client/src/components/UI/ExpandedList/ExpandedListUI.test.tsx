import React from 'react';
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import ExpandedListUI from './ExpandedListUI';

describe('ExpandedListUI TESTS', () => {
    test('ellipsed ExpandedListUI should not be changed', () => {
        render(
            <ExpandedListUI toggleButtonTexts={['Развернуть', 'Свернуть']}>
                <p>Вы любите смотреть фильмы онлайн и проводите много времени, прочесывая сайты в поисках чего-нибудь интересного? Стоит задержаться на ivi – фильмов, которые собраны у нас, вам хватит надолго. Коллекция постоянно пополняется как новыми фильмами, так и признанными шедеврами прошлых лет! Независимо от того, кто вы – любитель энергичных боевиков или поклонница молодежных сериалов, изобилие нашего каталога заставит вас забыть обо всех других способах проведения досуга, и вы будете пересматривать любимые фильмы онлайн снова и снова!</p>
                <p>Выбор фильмов очень широк и многообразен, так что каждый найдет для себя что-то интересное, каким бы ни были его вкусы. Предпочитаете картины исключительно зарубежного производства? У нас их предостаточно: это и золотая классика Голливуда, и душевные французские комедии, и темпераментные итальянские драмы, и шумные индийские музыкальные фильмы. А может, вы патриот и любите российские фильмы? Что ж, и таких фильмов у нас немало. Что вам больше по вкусу – добрая старая классика или новинки кинопроката? Неважно, каким будет ваш ответ – у нас есть все, как картины эпохи зарождения кинематографа, так 2016 года и фильмы 2015.</p>
            </ExpandedListUI>
        )
        const expandedList = screen.getByRole('expanded-wrapper');

        expect(expandedList).toMatchSnapshot()
    })

    test('ExpandedList should not be collapsed', () => {
        const spanTexts = ['Читать далее', 'Свернуть'];
        render(
            <ExpandedListUI 
                toggleButtonTexts={spanTexts}
                clipSize={3}
                clipFormat='horizontal'>
                <p>Вы любите смотреть фильмы онлайн и проводите много времени, прочесывая сайты в поисках чего-нибудь интересного? Стоит задержаться на ivi – фильмов, которые собраны у нас, вам хватит надолго. Коллекция постоянно пополняется как новыми фильмами, так и признанными шедеврами прошлых лет! Независимо от того, кто вы – любитель энергичных боевиков или поклонница молодежных сериалов, изобилие нашего каталога заставит вас забыть обо всех других способах проведения досуга, и вы будете пересматривать любимые фильмы онлайн снова и снова!</p>
                <p>Выбор фильмов очень широк и многообразен, так что каждый найдет для себя что-то интересное, каким бы ни были его вкусы. Предпочитаете картины исключительно зарубежного производства? У нас их предостаточно: это и золотая классика Голливуда, и душевные французские комедии, и темпераментные итальянские драмы, и шумные индийские музыкальные фильмы. А может, вы патриот и любите российские фильмы? Что ж, и таких фильмов у нас немало. Что вам больше по вкусу – добрая старая классика или новинки кинопроката? Неважно, каким будет ваш ответ – у нас есть все, как картины эпохи зарождения кинематографа, так 2016 года и фильмы 2015.</p>
            </ExpandedListUI>
        )
        const expandedList = screen.getByTestId('clipped-text');
        const expandButton = screen.getByText(spanTexts[0]);
        userEvent.click(expandButton);
        
        expect(expandedList).not.toHaveClass('is-truncated');
    })
})