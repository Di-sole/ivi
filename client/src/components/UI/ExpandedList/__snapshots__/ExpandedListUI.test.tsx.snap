// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandedListUI TESTS ellipsed ExpandedListUI should not be changed 1`] = `
<div
  class="jsx-812996627 clause"
  role="expanded-wrapper"
>
  <div
    class="jsx-812996627 clause-text is-truncated"
    data-testid="clipped-text"
    id="text-wrapper"
  >
    <div
      class="jsx-812996627 clause-text-inner hidden-children clause-clamped"
    >
      <p>
        Вы любите смотреть фильмы онлайн и проводите много времени, прочесывая сайты в поисках чего-нибудь интересного? Стоит задержаться на ivi – фильмов, которые собраны у нас, вам хватит надолго. Коллекция постоянно пополняется как новыми фильмами, так и признанными шедеврами прошлых лет! Независимо от того, кто вы – любитель энергичных боевиков или поклонница молодежных сериалов, изобилие нашего каталога заставит вас забыть обо всех других способах проведения досуга, и вы будете пересматривать любимые фильмы онлайн снова и снова!
      </p>
      <p>
        Выбор фильмов очень широк и многообразен, так что каждый найдет для себя что-то интересное, каким бы ни были его вкусы. Предпочитаете картины исключительно зарубежного производства? У нас их предостаточно: это и золотая классика Голливуда, и душевные французские комедии, и темпераментные итальянские драмы, и шумные индийские музыкальные фильмы. А может, вы патриот и любите российские фильмы? Что ж, и таких фильмов у нас немало. Что вам больше по вкусу – добрая старая классика или новинки кинопроката? Неважно, каким будет ваш ответ – у нас есть все, как картины эпохи зарождения кинематографа, так 2016 года и фильмы 2015.
      </p>
    </div>
    <span
      class="jsx-812996627 clause-toggle"
    >
      Развернуть
    </span>
  </div>
</div>
`;
